processing/src/main/java/org/apache/druid/segment/join/JoinableFactoryWrapper.java

```diff
@@ -178,15 +183,37 @@ static JoinClauseToFilterConversion convertJoinToFilter(
         }
 
         numValues -= columnValuesWithUniqueFlag.getColumnValues().size();
-        filters.add(Filters.toFilter(new InDimFilter(leftColumn, columnValuesWithUniqueFlag.getColumnValues())));
+        // For each column value which are received in order we increment the index
+        // and add it in the appropriate index in the map
+        int c = 0;
+        for (String val : columnValuesWithUniqueFlag.getColumnValues()) {
```

Comment:
The logic here is inefficient since it creates `filterMap` full of `SelectorDimFilter` even if it ends up not being used (in the case where `clause.getCondition().getEquiConditions().size() == 1`). We need to be careful about this stuff, since the filters can be quite large.

Also I am not sure if it's correct. It looks like this is doing a zip between all `clause.getJoinable().getNonNullColumnValues(condition.getRightColumn(), numValues)` for each equi-join condition. However, because `getNonNullColumnValues` returns _unique_ values, the different column values might not line up. I think if you add a test where the column values aren't all unique, you'll see this effect.

IMO, it's OK to fix this bug by limiting the join-to-filter optimization to the case where there is a single equi-join. It's not clear that the optimization is actually helpful when there's more than one equi-join, since we have to use a bunch of selectors with a bunch of boolean operations, rather than an "in" filter. So, I'd suggest doing that.

Reply:
Yes, agree with Gian's comment. Just thinking, can there be another solution to have the `in` filters but also retain the join incase `clause.getCondition().getEquiConditions().size() > 1`? That way we could send some filtering down to the scan cursor and do the remaining one in the join cursor for correctness.

Reply:
I see, I introduced the orderedSet to ensure the values lineup in order, but using duplicate values I could see the issue. Without using the linkedSet if I use unique non-sorted values in the input they also do not line up as for the TreeSet creating the wrong filters. I agree that we should limit the optimization as you suggested

Reply:
@rohangarg I am unsure if we can do that, Currently what we do is pass something like `(A|B)&(C|D)` to the base filter. What we actually need is `(A&B)|(C&D)`. Considering these values `A=0,B=0,C=1,D=1` 
```
(A|B)&(C|D) = 0 which the base cursor ignored
(A&B)|(C&D) = 1 which we need but since the base cursor already ignored that we cannot get it in post process
```
I think we can get correctness with the way Gian has suggested here and then look back separately if this can be achieved somehow through an interplay of cursors and filters

Reply:
I think if we currently pass the filter as `C1 IN (A, B) AND C2 IN (C, D)`, what we actually need is `(C1, C2) IN ((A, C), (B, D))`.
So, maybe we could pass the in filters down since `(A, C)` and `(B,D)` both satisfy `C1 IN (A, B) AND C2 IN (C, D)` - the join condition could filter more after the in filters to extract only `(A, C)` and `(B, D)` and eliminate tuples like `(A, D)` and `(B, C)`.